lang/java/avro/src/test/java/org/apache/avro/TestSchemaValidation.java

```diff
@@ -15,183 +15,316 @@
+import static org.apache.avro.TestSchemas.LONG_ARRAY_SCHEMA;
+import static org.apache.avro.TestSchemas.LONG_LIST_RECORD;
+import static org.apache.avro.TestSchemas.LONG_MAP_SCHEMA;
+import static org.apache.avro.TestSchemas.LONG_SCHEMA;
+import static org.apache.avro.TestSchemas.LONG_UNION_SCHEMA;
+import static org.apache.avro.TestSchemas.NULL_SCHEMA;
+import static org.apache.avro.TestSchemas.STRING_INT_UNION_SCHEMA;
+import static org.apache.avro.TestSchemas.STRING_SCHEMA;
+import static org.apache.avro.TestSchemas.STRING_UNION_SCHEMA;
+import static org.apache.avro.TestSchemas.list;
+
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
 
-import org.apache.avro.reflect.ReflectData;
+import org.apache.avro.TestSchemas.ReaderWriter;
 import org.junit.Assert;
+import org.junit.Rule;
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 
 public class TestSchemaValidation {
 
-  SchemaValidatorBuilder builder = new SchemaValidatorBuilder();
+  @Rule
+  public ExpectedException expectedException = ExpectedException.none();
+  
+  /** Collection of reader/writer schema pair that are compatible. */
+  public static final List<ReaderWriter> COMPATIBLE_READER_WRITER_TEST_CASES = list(
+      new ReaderWriter(BOOLEAN_SCHEMA, BOOLEAN_SCHEMA),
 
-  Schema rec = SchemaBuilder.record("test.Rec").fields()
-      .name("a").type().intType().intDefault(1)
-      .name("b").type().longType().noDefault()
-      .endRecord();
+      new ReaderWriter(INT_SCHEMA, INT_SCHEMA),
 
-  Schema rec2 = SchemaBuilder.record("test.Rec").fields()
-      .name("a").type().intType().intDefault(1)
-      .name("b").type().longType().noDefault()
-      .name("c").type().intType().intDefault(0)
-      .endRecord();
+      new ReaderWriter(LONG_SCHEMA, INT_SCHEMA),
+      new ReaderWriter(LONG_SCHEMA, LONG_SCHEMA),
 
-  Schema rec3 = SchemaBuilder.record("test.Rec").fields()
-      .name("b").type().longType().noDefault()
-      .name("c").type().intType().intDefault(0)
-      .endRecord();
+      // Avro spec says INT/LONG can be promoted to FLOAT/DOUBLE.
+      // This is arguable as this causes a loss of precision.
+      new ReaderWriter(FLOAT_SCHEMA, INT_SCHEMA),
+      new ReaderWriter(FLOAT_SCHEMA, LONG_SCHEMA),
+      new ReaderWriter(DOUBLE_SCHEMA, LONG_SCHEMA),
 
-  Schema rec4 = SchemaBuilder.record("test.Rec").fields()
-      .name("b").type().longType().noDefault()
-      .name("c").type().intType().noDefault()
-      .endRecord();
+      new ReaderWriter(DOUBLE_SCHEMA, INT_SCHEMA),
+      new ReaderWriter(DOUBLE_SCHEMA, FLOAT_SCHEMA),
 
-  Schema rec5 = SchemaBuilder.record("test.Rec").fields()
-      .name("a").type().stringType().stringDefault("") // different type from original
-      .name("b").type().longType().noDefault()
-      .name("c").type().intType().intDefault(0)
-      .endRecord();
+      new ReaderWriter(STRING_SCHEMA, STRING_SCHEMA),
 
-  @Test
```

Comment:
Several test cases are now removed, are those covered with the new testSchemaCompatibilitySuccesses and testSchemaCompatibilityFailures tests? In the deleted test cases I see validators where it tests for mutual read, this code path is not covered any more, is it?